l2-contracts/contracts/AttesterRegistry.sol

```diff
@@ -0,0 +1,159 @@
+        verifyExists(nodeOwner);
+        require(attesters[nodeOwner].isInactive, "Attester is still active");
+        require(epoch >= attesters[nodeOwner].inactiveSince + INACTIVITY_DELAY,
+            "Attester's inactivity delay has not passed"
+        );
+
+        // Remove from mapping.
+        delete attesters[nodeOwner];
+
+        // Remove from array by swapping the last element (gas-efficient, not preserving order).
+        attesterOwners[nodeOwnerIndex(nodeOwner)] = attesterOwners[attesterOwners.length - 1];
+        attesterOwners.pop();
+    }
+
+    // Inactivates an attester.
+    function inactivate(address nodeOwner) external onlyOwner {
+        verifyExists(nodeOwner);
+        attesters[nodeOwner].isInactive = true;
+        attesters[nodeOwner].inactiveSince = epoch;
+    }
+
+    // Activates an attester.
+    function activate(address nodeOwner) external onlyOwner {
+        verifyExists(nodeOwner);
+        attesters[nodeOwner].isInactive = false;
+        attesters[nodeOwner].inactiveSince = 0;
+    }
+
+    // Changes the weight.
+    function changeWeight(address nodeOwner, uint256 weight) external onlyOwner {
+        verifyExists(nodeOwner);
+        attesters[nodeOwner].weight = weight;
+    }
+
+    // Changes the public key.
+    function changePublicKey(address nodeOwner, bytes calldata pubKey) external onlyOwner {
+        verifyExists(nodeOwner);
+        attesters[nodeOwner].pubKey = pubKey;
+    }
+
+    // Creates a new committee list which will become the next committee list.
+    function setNextCommittee() external onlyOwner {
+        epoch += 1;
+
+        // Replace current committee with next committee.
+        delete committee;
+        for (uint256 i = 0; i < nextCommittee.length; i++) {
+            committee.push(nextCommittee[i]);
+        }
+
+        // Populate `nextCommittee` based on active attesters.
+        delete nextCommittee;
+        for (uint256 i = 0; i < attesterOwners.length; i++) {
+            Attester memory attester = attesters[attesterOwners[i]];
+            if (!attester.isInactive) {
+                nextCommittee.push(CommitteeAttester(attesterOwners[i], attester.weight, attester.pubKey));
+            }
+        }
+    }
+
+    // Finds the index of a node owner in the `nodeOwners` array.
+    function nodeOwnerIndex(address nodeOwner) private view returns (uint256) {
+        for (uint256 i = 0; i < attesterOwners.length; i++) {
+            if (attesterOwners[i] == nodeOwner) {
+                return i;
+            }
+        }
+        revert("nodeOwner not found");
+    }
+    // Verifies that an attester exists.
+    function verifyExists(address nodeOwner) private view {
+        require(attesters[nodeOwner].pubKey.length != 0, "Attester doesn't exist");
+    }
+
+    function compareBytes(bytes storage a, bytes calldata b) private pure returns (bool) {
+        return keccak256(a) == keccak256(b);
```

Comment:
update `keccak256(b)` to use the efficient call library to save from recopying the bytes array

Reply:
`keccak256(b)` already acts on `bytes calldata` and doesn't copy it into memory, no? Or did you mean something else?

Reply:
under the hood it will re-copy the array, with `EfficientCall.keccak` we modify the vm's fat pointer to point to the calldata slice

Reply:
The implementation is under L1 contracts. Is it possible to import it?

Reply:
if its in the l2 contracts you should be able to import the zksync contracts like [here](https://github.com/matter-labs/era-contracts/blob/b5ccc20cffa36f1395360d8762f1762f7409835f/l2-contracts/contracts/L2ContractHelper.sol#L5). if its an l1 contract then you wont be able to use it

Reply:
Thanks, it worked after I added the package to `devDependencies`, like in the branch you linked to. 

But `EfficientCall.keccak` supports only `calldata` types, resulting in the following: 
```solidity
  function compareBytes(bytes storage a, bytes calldata b) private view returns (bool) {
      return keccak256(a) == EfficientCall.keccak(b);
  }
```
WDYT?

Reply:
yea thats fine, itll at least cutdown on the gas for the second keccak call

Reply:
Added here: https://github.com/matter-labs/era-contracts/pull/555/commits/31d42cb8e3922e471e89988f633b067e6128f38a